src/Processors/QueryPlan/Optimizations/joinRuntimeFilter.cpp

```diff
@@ -0,0 +1,220 @@
+#include <memory>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/BuildRuntimeFilterStep.h>
+#include <Processors/QueryPlan/JoinStepLogical.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/Optimizations/Utils.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionsLogical.h>
+#include <Functions/IFunctionAdaptors.h>
+#include <Functions/FunctionFactory.h>
+#include <Interpreters/ActionsDAG.h>
+#include <Interpreters/Context.h>
+#include <Core/ColumnWithTypeAndName.h>
+#include <Core/Settings.h>
+#include <Common/thread_local_rng.h>
+#include <Common/logger_useful.h>
+#include <DataTypes/getLeastSupertype.h>
+#include <fmt/format.h>
+
+
+namespace DB
+{
+
+namespace QueryPlanOptimizations
+{
+
+const ActionsDAG::Node & createRuntimeFilterCondition(ActionsDAG & actions_dag, const String & filter_name, const ColumnWithTypeAndName & key_column, const DataTypePtr & filter_element_type)
+{
+    const auto & filter_name_node = actions_dag.addColumn(
+        ColumnWithTypeAndName(
+            DataTypeString().createColumnConst(0, filter_name),
+            std::make_shared<DataTypeString>(),
+            filter_name));
+
+    const auto & key_column_node = actions_dag. findInOutputs(key_column.name);
+    const auto * filter_argument = &key_column_node;
+
+    /// Cast to the type of filter element if needed
+    if (!key_column.type->equals(*filter_element_type))
+        filter_argument = &actions_dag.addCast(key_column_node, filter_element_type, {});
+
+    auto filter_function = FunctionFactory::instance().get("__filterContains", /*query_context*/nullptr);
+    const auto & condition = actions_dag.addFunction(filter_function, {&filter_name_node, filter_argument}, {});
+    return condition;
+}
+
+bool tryAddJoinRuntimeFilter(QueryPlan::Node & node, QueryPlan::Nodes & nodes, const QueryPlanOptimizationSettings & optimization_settings)
+{
+    /// Is this a join step?
+    auto * join_step = typeid_cast<JoinStepLogical *>(node.step.get());
+    if (!join_step)
+        return false;
+
+    /// Joining two sources?
+    if (node.children.size() != 2)
+        return false;
+
+    /// Check if join can do runtime filtering on left table
+    const auto & join_info = join_step->getJoinInfo();
+    const bool can_use_runtime_filter =
+        (
+            (join_info.kind == JoinKind::Inner && (join_info.strictness == JoinStrictness::All || join_info.strictness != JoinStrictness::Any)) ||
+            ((join_info.kind == JoinKind::Left || join_info.kind == JoinKind::Right) && join_info.strictness == JoinStrictness::Semi)
+        ) &&
```

Comment:
`INNER ANY` must be allowed as well because it filters the same rows from the left table, the only difference is the number of rows in the output. For `INNER`, you should forbid only `Asof`.

`ANTI JOIN` -- for now can be forbidden, but it can be supported with some LRU cache filter.